### ISSUE TITLE
Time interval is not behaving correctly

### ISSUE THREAD

**Issue by lasers** [ultrabug/py3status#1238]
I'm an idiot. In #1225, @ultrabug asked me to open an issue so I opened an issue https://github.com/ultrabug/py3status/issues/1235 about that. I later realize he wanted me to open an issue about the time.

Copypasta.

>I switched to `self.py3.time_in(0)` here to keep it simple.
>
>I'm not sure if I mentioned this before... I think the timing is off by a second as the correct code should be `self.py3.time_in(1)` to update every `1` second. You can try this PR with `1` and will get `pomodoro` to count down every `2` seconds instead.
>
>Maybe we should hold this off to look into this as everything is now off by a second+.

I assume this is reasonable.
```
self.py3.time_in(-1)   # update once, stop forever
self.py3.time_in(0)    # update twice every second
self.py3.time_in(1)    # update every 1 second
self.py3.time_in(5)    # update every 5 second
```
This is noticeable in audio-playing modules using timed placeholders too.

**Comment by tobes** [comment]
Actually things work as intended but maybe it is not clear.  The documentation may need to be improved or perhaps we change the behaviour to fit more with how people assume it works.

The essential thing is that `self.py3.time_in(1)` does not give you the time in 1 second.  It gives you the time in one second at the next second boundary. This is going to be in ~2 seconds time.

This is because we group updates together on time boundaries.

For something you want to run at the next second `time_in(sync_to=1)` is probably what you want.  If you after an off second time then `offset` is useful so that you update on the second of the thing rather than the second of time (if that makes any sense)

**Comment by tobes** [comment]
It may be that something like

```diffdiff --git a/py3status/py3.py b/py3status/py3.py
index eb0cdf53..1a8a5518 100644
--- a/py3status/py3.py
+++ b/py3status/py3.py
@@ -592,6 +592,9 @@ class Py3:
             else:
                 sync_to = 1
 
+        if seconds:
+            seconds = seconds - 0.1
+
         requested = time() + seconds - offset
 
         # if sync_to then we find the sync time for the requested time
```

would help as this makes the behaviour appear more as the user might expect. so that `time_in(1)` will likely be the time on the next second boundary ie in ~1 second

**Comment by tobes** [comment]
> self.py3.time_in(0)    # update twice every second

`self.py3.time_in(0.5)` should do that.

`self.py3.time_in(0)` should be updating as soon as possible (but on a second)

```diff
diff --git a/py3status/py3.py b/py3status/py3.py
index eb0cdf53..3d3a88f5 100644
--- a/py3status/py3.py
+++ b/py3status/py3.py
@@ -588,10 +588,16 @@ class Py3:
         # Unless the requested update is in less than a second
         if sync_to is None:
             if seconds and seconds < 1:
-                sync_to = 0
+                if 1 % seconds == 0:
+                    sync_to = seconds
+                else:
+                    sync_to = 0
             else:
                 sync_to = 1
```

Improves our syncing with partial seconds eg `time_in(0.5)` - will be on the half second boundary

**Comment by lasers** [comment]
Can confirm `time_in(sync_to=1)` works okay with `pomodoro`. Same as current `time_in(0)`.

When we use timed placeholders, that's where the problem usually is.... and that most modules does not have access to `sync=1` thing from the config.

That first diff does what I'm hoping it to... Using `5` gives me `5` interval... `pomodoro` winds down more gracefully (is nice) eg `1500, 1455, 1450, 1445, 1440...`

Without the second diff, I fluid between 2-3 updates per second... so I think we should add this too for better partial seconds. It also feels like right on time... tick...tock...tick...tock.

Few Q: For timers, should we use `1` over `0` or is it just perception (now with minus 0.1)? Which would `pomodoro` benefits greatly? `0`, `1`, or `sync_to=1`. There is `offset` too, but only `uptime` uses it at this time... Make a pull request to close this issue too?

**Comment by tobes** [comment]
> When we use timed placeholders

what does this mean?

> That first diff does what I'm hoping it to... Using 5 gives me 5 interval.

It may seem right but it is going to be wrong because for something like pomodoro time is really important.  For most modules time is not important.  Basically anything that displays a time is fundamentally different than some thing that just an update every so often.  Because for the user the exact time is vital.  So things like uptime, clock etc

> For timers, should we use 1 over 0 or is it just perception

For things that matter to the user you should do it properly which is why we have the `sync_to` and `offset` options, because they are exactly for this type of situation.  TBH the timing stuff is hard to get right unless you understand exactly what you are trying to do.  For these cases just using `time_in(x)` will usually be wrong.

>  There is offset too, but only uptime uses it at this time

`timer` module also uses it as it cares about the time.

>  Make a pull request to close this issue too?

I will, when I have had a chance to consider other options like `time_in(2.5)`  Also really the main thing is about improving the documentation as currently we can do most things if we use the function correctly.

**Comment by lasers** [comment]
>>When we use timed placeholders
>what does this mean?

Sorry. I meant time-related placeholders. Length, duration, elapsed, etc... 

>For things that matter to the user you should do it properly which is why we have the sync_to and offset options, because they are exactly for this type of situation.

Okay. I read you. Put out a fix for `pomodoro` and other modules too if need. I can test them next if need. I think I ran into this issue with `rate_counter` too so it could need that too.

Since it is just me that complained about the differentiated time intervals, maybe the fix is not needed at all... to keep it as-is (more delays) until somebody else come along with a complaint? Because of "TBH the timing stuff is hard to get right" and "just using time_in(x) will usually be wrong". Could do case-by-case basis for each module to use `sync_to` or such. Gn.

**Comment by lasers** [comment]
Closed via #1650.
